File: 2020/day_05/day_05.py
with open('in') as f:
    data = list(map(lambda x: x.strip(), f.readlines()))


def code_to_int(code):
    low, high = 0, 2 ** len(code) - 1
    for c in code:
        if c in 'FL':
            high = low + (high - low) // 2
        else:
            low = high - (high - low) // 2

    assert low == high
    return low


def part_one():
    max_id = 0
    for seat in data:
        row = code_to_int(seat[:7])
        column = code_to_int(seat[-3:])

        seat_id = row * 8 + column
        max_id = max(max_id, seat_id)

    return max_id


def part_two():
    existing_ids = set()
    for seat in data:
        row = code_to_int(seat[:7])
        column = code_to_int(seat[-3:])

        seat_id = row * 8 + column
        existing_ids.add(seat_id)

    for seat_id in range(0, 128 * 8):
        if seat_id not in existing_ids and seat_id - 1 in existing_ids and seat_id + 1 in existing_ids:
            return seat_id


if __name__ == '__main__':
    print(f'Part One: {part_one()}')
    print(f'Part Two: {part_two()}')

File: 2020/day_05/day_05.rs
use std::fs;

fn main() {
    let f = fs::read_to_string("in").unwrap();
    let data: Vec<&str> = f.lines().map(|x| x.trim()).collect();
    let ids: Vec<u32> = data.iter()
        .map(|x| (&x[0..7], &x[7..]))
        .map(|(x, y)| (x.replace("F", "0").replace("B", "1"), y.replace("L", "0").replace("R", "1")))
        .map(|(x, y)| (u32::from_str_radix(&x, 2).unwrap(), u32::from_str_radix(&y, 2).unwrap()))
        .map(|(x, y)| x * 8 + y).collect();

    println!("Part One: {}", part_one(&ids));
    println!("Part Two: {}", part_two(&ids));
}

fn part_one(data: &Vec<u32>) -> u32 {
    *data.iter().max().unwrap()
}

fn part_two(data: &Vec<u32>) -> u32 {
    let mut sorted = data.clone();
    sorted.sort();

    let (_, x) = sorted.iter().enumerate().skip(1).find(|(i, &x)| x - sorted[i-1] == 2).unwrap();
    *x-1
}